llvm/lib/IR/RepoTicket.cpp

```diff
@@ -74,54 +74,77 @@ updateInitialDigestAndGetDependencies(const GlobalObject *GO, MD5 &GOHash,
   GOInfoMap::const_iterator Pos;
   if (const auto GV = dyn_cast<GlobalVariable>(GO)) {
     ++GONum.VarNum;
-    Pos = calculateInitialDigestAndDependencies(GV, GOIMap);
+    Pos = calculateInitialDigestAndDependenciesAndContributions(GV, GOIMap);
   } else if (const auto Fn = dyn_cast<Function>(GO)) {
     ++GONum.FuncNum;
-    Pos = calculateInitialDigestAndDependencies(Fn, GOIMap);
+    Pos = calculateInitialDigestAndDependenciesAndContributions(Fn, GOIMap);
   } else {
     llvm_unreachable("Unknown global object type!");
   }
   GOHash.update(Pos->second.InitialDigest.Bytes);
   return Pos->second.Dependencies;
 }
 
-const DependenciesType &
-updateDigestGONumAndGetDependencies(const GlobalObject *GO, MD5 &GOHash,
-                                    GOInfoMap &GOIMap, GONumber &GONum) {
+const GODigestState updateDigestGONumAndGetDependencies(const GlobalObject *GO,
+                                                        MD5 &GOHash,
+                                                        GOInfoMap &GOIMap,
+                                                        GONumber &GONum) {
   auto It = GOIMap.find(GO);
   if (It != GOIMap.end()) {
     const GOInfo &GOInformation = It->second;
     GOHash.update(GOInformation.InitialDigest.Bytes);
-    return GOInformation.Dependencies;
+    return {GOInformation.Dependencies, true};
   }
 
-  return updateInitialDigestAndGetDependencies(GO, GOHash, GOIMap, GONum);
+  return {updateInitialDigestAndGetDependencies(GO, GOHash, GOIMap, GONum),
+          true};
 }
 
-const DependenciesType &updateDigestAndGetDependencies(const GlobalObject *GO,
-                                                       MD5 &GOHash,
-                                                       GOInfoMap &GOIMap) {
+const GODigestState updateGODigestAndGetDependencies(const GlobalObject *GO,
+                                                     MD5 &GOHash,
+                                                     GOInfoMap &GOIMap) {
   if (auto GOMD = GO->getMetadata(LLVMContext::MD_repo_ticket)) {
     if (const TicketNode *TN = dyn_cast<TicketNode>(GOMD)) {
       DigestType D = TN->getDigest();
       GOHash.update(D.Bytes);
-      return GOIMap.try_emplace(GO, std::move(D), DependenciesType())
-          .first->second.Dependencies;
+      return {GOIMap
+                  .try_emplace(GO, std::move(D), DependenciesType(),
+                               ContributionsType())
+                  .first->second.Dependencies,
+              true};
     }
     report_fatal_error("Failed to get TicketNode metadata!");
   }
   GONumber GONum;
-  return updateInitialDigestAndGetDependencies(GO, GOHash, GOIMap, GONum);
+  return {updateInitialDigestAndGetDependencies(GO, GOHash, GOIMap, GONum),
+          true};
+}
+
+const GODigestState updateDigestUseMemoryDependencies(const GlobalObject *GO,
```

Comment:
Are "memory dependencies" the same as "contributions"? If so, we probably shouldn't introduce a second term!

Reply:
Yes, I have changed it to "contributions".